tests/integration/test_charm.py

```diff
@@ -5,19 +5,104 @@
 import logging
 
 import pytest
+from helpers import (
+    check_prometheus_is_ready,
+    get_job_config_for,
+    get_prometheus_config,
+    get_prometheus_rules,
+    get_rules_for,
+)
 
 logger = logging.getLogger(__name__)
 
 
 @pytest.mark.abort_on_fail
-async def test_build_and_deploy_with_ubuntu_image(ops_test):
+@pytest.mark.parametrize("oci_image", ["ubuntu/prometheus:latest", "prom/prometheus:latest"])
+async def test_build_and_deploy_with_alternative_images(ops_test, prometheus_charm, oci_image):
     """Test that the Prometheus charm can be deployed successfully."""
-    prometheus_charm = await ops_test.build_charm(".")
-    resources = {"prometheus-image": "ubuntu/prometheus:latest"}
+    resources = {"prometheus-image": oci_image}
+    app_name = "prometheus-" + oci_image.split("/")[0]
 
+    await ops_test.model.deploy(prometheus_charm, resources=resources, application_name=app_name)
+    await ops_test.model.wait_for_idle(apps=[app_name], status="active")
+    await ops_test.model.block_until(lambda: len(ops_test.model.applications[app_name].units) > 0)
```

Comment:
Isn't this already captured by `wait_for_idle(..., status="active")`?

Reply:
As far as I am aware it is not.

Reply:
Hmm, this does seem like a strange ordering -- we first check that the app is `active/idle`, then block on number of units being > 0? Surely this will never block?

Reply:
I do find that reversing the ordering or removing the `block_until` requested units are up makes the tests flaky.

Reply:
App status is supposed to be set automatically to the " worst across all units".
Any chance charm code is touchig app status? Or maybe charm code sets unit status to active prematurely?

Reply:
The Prometheus charm only sets `ActiveStatus` for unit status (see `charm.py`). It does this either after reloading the prometheus configuration or setting a new pebble layer. I do not see how this can be considered "premature". Further these integrations tests (as yet) deploy only a single unit.

Reply:
Looking at `_configure()` it seems that it may be possible that a unit status is set to active before startup sequence is complete. Not sure. For example:
1. config changed - > pebble ready -> (charm code changes status to active) -> (something else, e.g. leadership event) -> start
1. pebble ready -> (charm code changes status to active) -> config changed -> start
1. a subsequent call to `_configure` flips status to Blocked because `_is_valid_timespec` failed for some reason?

Reply:
Not sure what you are trying to say here.
1. If pebble is not ready when config changed executes `_configure()` returns from `can_connect()`. There is no use case for leadership or start here.
2. If pebble ready executes before config changed then a Prometheus configuration will be pushed and workload will be restarted or reloaded. So unit is marked as active which is valid.
3. If at any point Prometheus receives an invalid configuration it will set its status to blocked. This is as desired. This is certainly not the case in this the integration test in this review as may be easily checked by logs and even dumping the config.

Reply:
If case (2) is possible (haven't been paying attention if in recent juju versions pebble ready can fire before config changed), then it can:
`pebble ready -> (charm code changes status to active) -> config changed (status is automatically changed to maintenance or something else) -> _configure changes status back to active -> start`
However, [default idle period](https://github.com/juju/python-libjuju/blob/c68ee681995dc4b966ed14160d61a72f249c3b7b/juju/model.py#L2554) is 15 sec so that is probably not the case. :thinking:

Reply:
I do think however that if `await ops_test.model.wait_for_idle(apps=[app_name], status="active")` is not sufficient then your integration test found a bug in charm code. (Congratulations! :stuck_out_tongue_closed_eyes: )

Reply:
Again I am totally lost on what you are trying to say. Both pebble ready and config changed events are handled by the same method `_configure()`. If pebble ready runs before config changed then this method ensures that the workload is indeed "activated". When config changed runs after pebble ready, if nothing else has changed in the mean time the config changed method has no effect.

Reply:
> I do think however that if `await ops_test.model.wait_for_idle(apps=[app_name], status="active")` is not sufficient then your integration test found a bug in charm code. (Congratulations! stuck_out_tongue_closed_eyes )

Can you explain why
1. It is guaranteed that `ops_test.model.wait_for_idle` will ensure that the required number of units are deployed ?
2. Even if `ops_test.model.wait_for_idle` does guarantee that required number of units are deployed, why should blocking on the required number of units after that make any difference ?
3. How does this demonstrate any bug in the charm code ?

Reply:
Watching `juju status` during my deployments I see the app status only settles to idle-active when the startup sequence is complete. If units are being brought up I think the status is "executing" or "maintenance". So wait_for_idle, in this context, should _indirectly_ guarantee that all units are up. The wait period (default of 15 sec) is used to makes sure it is not a fluke.
afaiu, If you are getting idle-active for 15 sec before the unit is up then it may be a bug in charm code. Feel free to bring a counter example.

Reply:
How does this suggest it is a bug in the Prometheus charm code rather than a misinterpretation of the Python libjuju API or a bug there or in Juju or pytest-operator ?

Reply:
It looks like there may be a bug in python-libjuju
https://github.com/juju/python-libjuju/issues/574

Regardless, I would move the `block_until` above `wait_for_idle`, in which case the python-libjuju bug should not have any effect.

Reply:
btw, since this is an integration tests PR, imo it can be merged if the tests are good, even if they are failing due to a _charm_ bug.